Sources/AWSSDKSwiftCore/Credential/CredentialProvider.swift

```diff
@@ -0,0 +1,61 @@
+//
+//  CredentialProvider.swift
+//  AWSSDKSwift
+//
+//  Created by Adam Fowler on 2019/12/5
+//
+//
+import AWSSigner
+import Foundation
+import NIO
+import NIOConcurrencyHelpers
+
+/// Protocol providing future holding a credential
+protocol CredentialProvider {
+    mutating func getCredential() -> EventLoopFuture<Credential>
```

Comment:
I understand, that we have need the `mutating` aspect in order to allow `struct`s. I'm not a big fan though. Since the credentialProvider is just initialized once, the benefits of a `struct` vs a `class` are IMHO neglectable.

Reply:
that was left from me trying something else. removed